File: backend/app/schemas/asr.py
# Copyright (c) 2026 Weave Thinker Contributors
# SPDX-License-Identifier: Apache-2.0

from pydantic import BaseModel
from typing import Optional, List, Dict, Any


class HotwordItem(BaseModel):
    text: str
    weight: int
    lang: Optional[str] = None


class HotwordListResponse(BaseModel):
    hotwords: List[HotwordItem]


class HotwordListRequest(BaseModel):
    hotwords: List[HotwordItem]


class ASRRequest(BaseModel):
    language: Optional[str] = "auto"
    format: Optional[str] = "wav"
    return_timestamps: Optional[bool] = True
    use_hotwords: Optional[bool] = True
    custom_hotwords: Optional[List[HotwordItem]] = None
    context: Optional[str] = None


class SegmentInfo(BaseModel):
    speaker: str
    speaker_confidence: Optional[float] = None
    start_time: float
    end_time: float
    text: str


class TimestampInfo(BaseModel):
    start_time: float
    end_time: float
    text: str


class ASRResponse(BaseModel):
    text: str
    language: Optional[str] = None
    timestamps: Optional[List[TimestampInfo]] = []
    segments: Optional[List[SegmentInfo]] = []
    hotwords_used: Optional[List[str]] = []
    speaker_mode: Optional[str] = "disabled"
    duration: Optional[float] = None

File: frontend/src/api/asr.ts
// Copyright (c) 2026 Weave Thinker Contributors
// SPDX-License-Identifier: Apache-2.0

import api from './client'
import { TOKEN_KEY } from './client'

export interface TimestampInfo {
  start_time: number
  end_time: number
  text: string
}

export interface SegmentInfo {
  speaker: string
  speaker_confidence: number | null
  start_time: number
  end_time: number
  text: string
}

export interface ASRResponse {
  text: string
  language?: string
  timestamps?: TimestampInfo[]
  segments?: SegmentInfo[]
  hotwords_used?: string[]
  speaker_mode?: string
  duration?: number
}

export interface HotwordItem {
  text: string
  weight: number
  lang?: string
}

export interface ASRStreamStartPayload {
  event: 'start'
  language?: string
  use_hotwords?: boolean
  custom_hotwords?: HotwordItem[]
  context?: string
  chunk_size_sec?: number
  unfixed_chunk_num?: number
  unfixed_token_num?: number
}

export interface ASRStreamEventPayload extends ASRResponse {
  event: 'ready' | 'partial' | 'segment' | 'final' | 'error'
  processed_seconds?: number | null
  total_seconds?: number | null
  chunk_index?: number
  final?: boolean
  backend?: string
  error?: string
}

export function getAsrWebSocketUrl(token = localStorage.getItem(TOKEN_KEY) || ''): string {
  const url = new URL('/api/asr/ws/transcribe/stream', window.location.origin)
  url.protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:'

  if (token) {
    url.searchParams.set('token', token)
  }

  return url.toString()
}

export const asrApi = {
  async transcribe(audioBlob: Blob, customHotwords?: HotwordItem[]): Promise<ASRResponse> {
    const formData = new FormData()
    formData.append('file', audioBlob, 'audio.wav')
    if (customHotwords && customHotwords.length > 0) {
      formData.append('custom_hotwords', JSON.stringify(customHotwords))
    }

    const { data } = await api.post<ASRResponse>('/asr/transcribe', formData, {
      headers: {
        'Content-Type': undefined
      }
    })

    return data
  },

  async getHotwords(): Promise<HotwordItem[]> {
    const { data } = await api.get<{ hotwords: HotwordItem[] }>('/asr/hotwords')
    return data.hotwords || []
  },

  async saveHotwords(hotwords: HotwordItem[]): Promise<HotwordItem[]> {
    const { data } = await api.post<{ hotwords: HotwordItem[] }>('/asr/hotwords', { hotwords })
    return data.hotwords || []
  }
}
